src/hotspot/share/oops/fieldInfo.inline.hpp

```diff
@@ -94,28 +105,33 @@ inline void Mapper<CON>::map_field_info(const FieldInfo& fi) {
 
 
 inline FieldInfoReader::FieldInfoReader(const Array<u1>* fi)
-  : _r(fi->data(), 0),
+  : _r(fi->data(), fi->length()),
     _next_index(0) { }
 
+inline void FieldInfoReader::read_field_counts(int *java_fields, int *injected_fields) {
+  *java_fields = _r.next_uint();
+  *injected_fields = _r.next_uint();
+}
+
 inline void FieldInfoReader::read_field_info(FieldInfo& fi) {
```

Comment:
I suggest you refactor this guy to use a new API point for the first two fields, `read_name_and_signature`.
Then you can reuse `read_name_and_signature` in the search comparator rather than use dead reckoning.